### PR TITLE
ci: integrate semgrep static analysis checks

### PR DESCRIPTION
**Pull request by pb-seiso**
# Contributor Comments
This adds the static analysis tool `semgrep` to run its default ruleset on non-test Python, in addition to the existing `bandit` security checks.

Currently, there are four findings related to use of `xml.etree` instead of `defusedxml` that are being suppressed. These are required for type hinting and cannot be removed until `defusedxml` supports it. See [this](https://github.com/tiran/defusedxml/issues/48) and [this](https://github.com/tiran/defusedxml/issues/54).

## Pull Request Checklist
Thank you for submitting a contribution to `easy_sast`!

In order to streamline the review of your contribution we ask that you review and comply with the below requirements:
 - [x] Ensure that `make build` completes successfully
 - [x] Keep our reports up-to-date by committing the output of `make report`
 - [x] Rebase your branch against the latest commit of the target branch, which likely should be `master`
 - [x] If there is an issue associated with your Pull Request, align your PR title with [this documentation](https://help.github.com/en/articles/closing-issues-using-keywords)

### CODE REVIEW

File: reports/semgrep_report.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+veracode/check_compliance.py
+severity:error rule:python.lang.security.use-defused-xml.use-defused-xml: Found use of the native Python XML libraries, which is vulnerable to XML external entity (XXE)
+attacks. The Python documentation recommends the 'defusedxml' library instead if the XML being
+loaded is untrusted.
+
+9:from xml.etree import (  # nosec (Used only when TYPE_CHECKING)
```
Comment (dlukeomalley) [inline reports/semgrep_report.json:6]: @pb-seiso Semgrep maintainer here 👋  

Did you see [Semgrep Docs > Managing Findings](https://semgrep.dev/docs/running-rules/#managing-findings) for details on Semgrep's `nosec`-like feature? I think it'll do what you want for suppression, and happy to open tickets on Semgrep if not!

Comment (pb-seiso) [reply reports/semgrep_report.json:6]: Thanks Luke! I missed that when reviewing the docs. That's exactly what I was looking for.

File: Dockerfile [hunk 2]
```diff
@@ -62,7 +62,8 @@ CMD ["run", "-m", "unittest", "discover", "-s", "tests", "-p", "test_*.py"]
 ## Security Tests
 FROM ci AS test_security
 ENTRYPOINT find . -type f -name '*.py' -exec bandit {} + && \
-  trufflehog --regex --entropy=False file:///usr/src/app/ --exclude_paths .truffleHog-exclude.txt
+  trufflehog --regex --entropy=False file:///usr/src/app/ --exclude_paths .truffleHog-exclude.txt && \
```
Comment (JonZeolla) [inline Dockerfile:65]: NIT: I have recently moved to the following approach which makes it easier to remove/change lines independently:
```bash
ENTRYPOINT echo "line 1" \
  && echo "line 2" \
  && echo "line 3"
```

File: Makefile [hunk 3]
```diff
@@ -160,4 +160,5 @@ report_coverage: test_unit
 .PHONY: report_security
 report_security: test_security
 	@python3 -c 'print("Updating the security testing reports...")'
-	@docker run --rm -v $$(pwd):/usr/src/app/ $(IMAGE_NAME):latest-test_security && find . -type f -name '*.py' -exec bandit --format json -o reports/bandit_report.json {} +
+	@docker run --rm -v $$(pwd):/usr/src/app/ $(IMAGE_NAME):latest-test_security && find . -type f -name '*.py' -exec bandit --format json -o reports/bandit_report.json {} + && \
```
Comment (JonZeolla) [inline Makefile:163]: Let's break this into 3 lines to stay consistent

**Review by JonZeolla — CHANGES_REQUESTED** [review]
`semgrep` appears to run but says that 1 file couldn't be analyzed; was that intentional?  Should we add `--strict`?
```bash
using config from https://semgrep.dev/p/r2c-ci. Visit https://semgrep.dev/registry to see all public rules.
downloading config...
excluding files:
- tests/*.py
running 119 rules...
ran 119 rules on 14 files: 0 findings
1 files could not be analyzed; run with --verbose for details or run with --strict to exit non-zero if any file cannot be analyzed
```

Also, when running `make report` I got the following error and I notice that the size of `semgrep_report.json` is 0.
```bash
[json]	INFO	JSON output written to file: reports/bandit_report.json
/bin/sh: semgrep: command not found
make: *** [report_security] Error 127
```

**Comment by pb-seiso** [comment]
@JonZeolla 

> `semgrep` appears to run but says that 1 file couldn't be analyzed; was that intentional? Should we add `--strict`?
> 
> ```shell
> using config from https://semgrep.dev/p/r2c-ci. Visit https://semgrep.dev/registry to see all public rules.
> downloading config...
> excluding files:
> - tests/*.py
> running 119 rules...
> ran 119 rules on 14 files: 0 findings
> 1 files could not be analyzed; run with --verbose for details or run with --strict to exit non-zero if any file cannot be analyzed
> ```

I've added both `--strict` and `--verbose` flags and added `reports/` to the exclude list, which takes care of the file that wasn't analyzed. 

> Also, when running `make report` I got the following error and I notice that the size of `semgrep_report.json` is 0.

`semgrep_report.json` will be empty if there are no findings, which should be the case currently (what I checked in).

> ```shell
> [json]	INFO	JSON output written to file: reports/bandit_report.json
> /bin/sh: semgrep: command not found
> make: *** [report_security] Error 127
> ```

This is due to missing `semgrep` locally. Two options are:

1. Document it as a dependency in the `README` along with `bandit`
2. Spin up a virtualenv and install `semgrep`, `bandit`, and other local dependencies there. I would create a separate PR for that before completing this PR.

Thoughts?

File: Makefile [hunk 4]
```diff
@@ -160,4 +160,6 @@ report_coverage: test_unit
 .PHONY: report_security
 report_security: test_security
 	@python3 -c 'print("Updating the security testing reports...")'
-	@docker run --rm -v $$(pwd):/usr/src/app/ $(IMAGE_NAME):latest-test_security && find . -type f -name '*.py' -exec bandit --format json -o reports/bandit_report.json {} +
+	@docker run --rm -v $$(pwd):/usr/src/app/ $(IMAGE_NAME):latest-test_security \
```
Comment (JonZeolla) [inline Makefile:163]: This is a mistake; we should be running the `find` and `semgrep` in the container.  `/bin/bash -c "x && y"`

**Review by JonZeolla — APPROVED** [review]

**Pull request merged**